### PR TITLE
(ignore) Experimental code about supporting PyTorch

### PR DESCRIPTION
**Pull request by Kelang-Tian**
Add PyTorch training loop to Elasticdl. Bugs need to be fixed
` File "../worker/my_worker_pytorch.py", line 482, in report_gradient
    return self.report_gradient_to_ps(grads)`

### CODE REVIEW

**Comment by skydoorkai** [comment]
1. Since this PR is for PS-only and training-only, better to remove all other codes (evaluation, allreduce, etc) to make it clean and easy to read. Also remove those comment-out codes.

1. Add comments where TensorFlow codes are used, starting why they are used and what will be the next step later.

**Comment by Kelang-Tian** [comment]
> * Since this PR is for PS-only and training-only, better to remove all other codes (evaluation, allreduce, etc) to make it clean and easy to read. Also remove those comment-out codes.
> * Add comments where TensorFlow codes are used, starting why they are used and what will be the next step later.

get it

**Comment by Kelang-Tian** [comment]
[This version](https://github.com/sql-machine-learning/elasticdl/pull/2233/commits/e792c867981baec81e5c917a4e2dc3efc799ab28) supports PyTorch `forward` function. `unittest` works after `grpc` bug fixed. 

**TODO List:**
1. [K.get_value(self._model.optimizer.lr)](https://github.com/sql-machine-learning/elasticdl/blob/e9e8c4b0af0b35a76baa9ee482ac29961a753ec8/elasticdl/python/worker/worker.py#L440) returns `lr` in `report_gradient_to_ps`, as PyTorch's optimizer does not have `.lr` attribute. Code needs to be modified.
2. Complete mnist train with `dataloader` needs `unittest`.
3. `evaluation` and `prediction` need to be completed.

File: elasticdl/python/worker/worker_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,684 @@
...
+    # TODO: Multiple tests are currently using this function to initialize
+    # self._model, where the initialization should be done via constructor.
+    def set_model(self, model_inst):
+        """Set model instance to worker."""
+        self._model = model_inst
+        self._model.float()
+        self._train_eagerly = False
+        # self._init_embeddings()
+
+    def _init_embedding_layer(self):
```
Comment (skydoorkai) [inline elasticdl/python/worker/worker_pytorch.py]: Is ElasticDL Embedding supported by PyTorch implementation? If not yet, remove embedding related codes.

Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: ok

File: elasticdl/python/tests/worker_ps_pytorch_test.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+                local_results.append(
+                    (float(loss.numpy()), float(acc_meter.result().numpy()))
+                )
+                acc_meter.reset_states()
+            if step > stop_step:
+                break
+        for w, l in zip(worker_results, local_results):
+            self.assertTupleEqual(w, l)
+
+    """
```
Comment (workingloong) [inline elasticdl/python/tests/worker_ps_pytorch_test.py]: It is better to remove unused codes.

File: elasticdl/python/worker/worker_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+    make CustomDataset for PyTorch
+    """
+    def __init__(self, data):
+        self.data_source = data
+
+    def __iter__(self):
+        return iter(self.data_source)
+
+
+class Worker_pytorch(object):
```
Comment (workingloong) [inline elasticdl/python/worker/worker_pytorch.py]: Worker_pytorch -> WorkerPytorch

File: elasticdl/python/tests/worker_ps_pytorch_test.py (hunk 2 above)
Comment (Kelang-Tian) [reply elasticdl/python/tests/worker_ps_pytorch_test.py]: There are some bugs need to be fixed. I will finish as soon.

File: elasticdl/python/worker/worker_pytorch.py (hunk 3 above)
Comment (Kelang-Tian) [reply elasticdl/python/worker/worker_pytorch.py]: Get it.

File: model_zoo/mnist/mnist_subclass_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        else:
+            return features, tf.cast(r["label"], tf.int32)
+
+    dataset = dataset.map(_parse_data)
+
+    if mode == Mode.TRAINING:
+        dataset = dataset.shuffle(buffer_size=1024)
+    return dataset
+
+def eval_metrics_fn():
```
Comment (skydoorkai) [inline model_zoo/mnist/mnist_subclass_pytorch.py:87]: Remove it as eval is not supported. And not with tf code.

File: model_zoo/mnist/mnist_subclass_pytorch.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+class CustomDataset(torch.utils.data.IterableDataset):
+    def __init__(self, data):
+        self.data_source = data
+
+    def __iter__(self):
+        return iter(self.data_source)
+
+def dataset_pytorch(dataset, minibatch_size):
```
Comment (skydoorkai) [inline model_zoo/mnist/mnist_subclass_pytorch.py]: Which dataset func is needed for pytorch model, `dataset_pytorch` or  `dataset_fn`?

Comment (Kelang-Tian) [reply model_zoo/mnist/mnist_subclass_pytorch.py]: At the moment, both functions required. `dataset_fn` makes tf dataset and `dataset_pytorch ` transform the data set from tf style to PyTorch format. 
 `dataset_fn`: recordIO data -> tf dataset
`dataset_pytorch `: tf dataset -> PyTorch dataset
This is not reasonable enough and will be modified later like `dataset_fn` : recordIO data -> PyTorch dataset

**Pull request closed without merging**